### ISSUE TITLE
LSRA throughput: Eliminate gtLsraInfo

### ISSUE THREAD

**Issue by CarolEidt** [dotnet/coreclr#7255]
Eliminate gtLsraInfo and combine the TreeNodeInfoInit with the building of RefPositions.

Some more info on this issue:
To Do: The current implementation actually includes five traversals. The interval building is actually done in three traversals, which should be merged into one: The first traversal (which is actually the second phase of Lowering) records the constraints and hints. The second traversal builds the intervals – this could be done with the first traversal. The third traversal sets the last use bits. This is done separately because there are currently inconsistencies between the gtNext/gtPrev links and the actual order of codegen. Once this is resolved, the lastUse bits should be set prior to register allocation by dataflow and/or SSA analysis. (Bugs 464609 and 750314)
 
The gtLsraInfo could/should be eliminated with the merging of the first two traversals – it is really only a mechanism for communicating information from the Lowering::TreeNodeInfoInit() phase to the LinearScan::buildIntervals() phase.

**Comment by mikedn** [comment]
@CarolEidt Any plans/timeline for this? Perhaps I could help with the implementation.

**Comment by CarolEidt** [comment]
@mikedn Thanks for asking! I've actually started working on this, albeit somewhat indirectly. The plan is to separate the `TreeNodeInfoInit` functionality into 1) determining which nodes are "contained", and thereby which nodes need registers, and 2) constraining the required registers. These could eventually be done in a single pre-post traversal, but I would probably start by leaving 1) in the current traversal, and incorporating 2) into the `RefPosition` building, eliminating `gtLsraInfo`. This requires (or at least is made cleaner by) a more formal representation of "containedness" which is what I'm working on now (and hope to send for review soon).

I hadn't really thought about splitting up the work, but the splitting of `TreeNodeInfoInit` could probably be done in parallel (and prior to the elimination of `gtLsraInfo). One might split it into something like `IdentifyContainedOperands`, `SetTargetRegConstraints` and `SetSourceRegConstraints'.

@sivarv @dotnet/jit-contrib - thoughts?

**Comment by mikedn** [comment]
> This requires (or at least is made cleaner by) a more formal representation of "containedness" which is what I'm working on now (and hope to send for review soon).

I was thinking that the "make contained" part could be moved from TreeNodeInfoInitX to LowerX and with register requirements setup moving to buildRefPositions TreeNodeInfoInit would simply disappear. As for representation of containment I wonder if we can't just pull contained nodes out of the linear order. Nodes having null gtNext/gtPrev would be considered contained and be reachable only through the nodes that directly use them.

> One might split it into something like IdentifyContainedOperands, SetTargetRegConstraints and SetSourceRegConstraints.

Well, I don't know, it always seemed to me that TreeNodeInfoInit does a lot more than it should and it should be split up somehow, otherwise this particular task would be nightmarish.

**Comment by pgavlin** [comment]
>  As for representation of containment I wonder if we can't just pull contained nodes out of the linear order. Nodes having null gtNext/gtPrev would be considered contained and be reachable only through the nodes that directly use them.

cc @russellhadley @sivarv 

We've discussed this on and off internally. This certainly has the nice effect of more clearly representing when a contained node executes (i.e. as part of its containing node). There are a few issues with contained nodes--especially the fact that inputs to the contained nodes become inputs to the containing node--but AFAICT we already handle these issues, so they may not be problematic.

**Comment by mikedn** [comment]
> There are a few issues with contained nodes--especially the fact that inputs to the contained nodes become inputs to the containing node

Hmm, right. At least they don't end up in the operand list of the container as in Phoenix :smile:.

**Comment by CarolEidt** [comment]
I am currently working on the following:
- Create a regreqts{arch}.cpp (e.g. regreqtsxarch.cpp) for each target.
- Move all the TreeNodeInfoInit methods out of lower{arch}.cpp into regreqts{arch}.cpp (this will leave only methods that are called during the first phase of Lowering).

Then I propose to (in some number of steps yet to be determined):
- Move non-register-requirements-related functionality out of the `TreeNodeInfoInit` methods into the first phase of `Lowering`.
- Modify all the `TreeNodeInfoInit` methods to be members of `LinearScan`, and to access the `gtLsraInfo` through access methods on `LinearScan` (at this point simply getting it from the node's `gtLsraInfo`).
- Remove the `gtLsraInfo` from the `GenTree` nodes, and maintain a mapping from `GenTree*` to `NodeInfo` for just the nodes currently in flight. Initially this will require maintaining the mapping for ranges of nodes, as we won't know which nodes actually produce a value until the ancestors have been visited.
- Finally, consider setting the containedness information in the single pass of `Lowering`. This will require some way to represent this, and I have some ideas that I'm experimenting with, but I think it's left as a later phase of this process.
@dotnet/jit-contrib @mikedn - thoughts?

**Comment by sivarv** [comment]
Proposed plan sounds good to me.

**Comment by mikedn** [comment]
> Create a regreqts{arch}.cpp (e.g. regreqtsxarch.cpp) for each target.

lsra{arch}.cpp? "regreqts" is unreadable.

> Move all the TreeNodeInfoInit methods out of lower{arch}.cpp into regreqts{arch}.cpp (this will leave only methods that are called during the first phase of Lowering).

That will conflict with my #8672 which is sort of by accident part of "Move non-register-requirements-related functionality out of the TreeNodeInfoInit methods into the first phase of Lowering"

> Remove the gtLsraInfo from the GenTree nodes, and maintain a mapping from GenTree* to NodeInfo for just the nodes currently in flight

Using a hashtable? I've tried something like that a long time ago and it seemed to be a regression in terms of performance. Though I never quite finished my experiment so perhaps there was some room for improvement.

**Comment by CarolEidt** [comment]
> lsra{arch}.cpp? "regreqts" is unreadable.

Sounds like a reasonable alternative to me.

>> Remove the gtLsraInfo from the GenTree nodes, and maintain a mapping from GenTree* to NodeInfo for just the nodes currently in flight

> Using a hashtable? I've tried something like that a long time ago and it seemed to be a regression in terms of performance. Though I never quite finished my experiment so perhaps there was some room for improvement.

There is already a hashtable that does essentially this - `operandToLocationInfoMap`. This would simply set and get the node info directly from there.

**Comment by pgavlin** [comment]
> Move non-register-requirements-related functionality out of the TreeNodeInfoInit methods into the first phase of Lowering.

When you do this, I would suggest leaving/moving any architecture-specific code in the `lower{arch}.cpp` files.

> lsra{arch}.cpp? "regreqts" is unreadable.

+1

> >Remove the gtLsraInfo from the GenTree nodes, and maintain a mapping from GenTree* to NodeInfo for just the nodes currently in flight
>
> Using a hashtable? I've tried something like that a long time ago and it seemed to be a regression in terms of performance. Though I never quite finished my experiment so perhaps there was some room for improvement.

If the number of nodes in flight is relatively small, the `SmallHashTable` should be pretty good: we get good performance using this type for a map from `GenTree*` to `LocationInfoList` when building ref positions, during which only a small number of mappings are needed at any given time.

SGTM otherwise.

**Comment by mikedn** [comment]
> If the number of nodes in flight is relatively small, the SmallHashTable should be pretty good: we get good performance using this type for a map from GenTree* to LocationInfoList when building ref positions, during which only a small number of mappings are needed at any given time.

Sounds good. My old experiment placed ALL the nodes in a SmallHashTable.

**Comment by CarolEidt** [comment]
>> Move non-register-requirements-related functionality out of the TreeNodeInfoInit methods into the first phase of Lowering.

> When you do this, I would suggest leaving/moving any architecture-specific code in the lower{arch}.cpp files.

I'm not sure I understand what you're suggesting. I think that the architecture-specific code is the precise reason to have the (newly-proposed-name) lsra{arch}.cpp files. Defining register requirements is quintessentially architecture-specific. However, the proposal is to separate the `Lowering` functionality from the register requirements specifications. So we have:
- `Lowering` which modifies the trees, possibly creating new nodes and new blocks, and ensures that the IR is in a form amenable to register allocation and code generation. The architecture-specific elements of this process would remain in lower{arch}.cpp.
- `TreeNodeInfoInit` which specifies the register requirements for each node. The architecture-specific elements of this would be in lsra{arch}.cpp.

**Comment by pgavlin** [comment]
> The architecture-specific elements of this process would remain in lower{arch}.cpp.

This is exactly what I was suggesting. I simply wanted to make sure that we don't move any (more) architecture-specific code into lower.cpp.

**Comment by CarolEidt** [comment]
> That will conflict with my #8672 which is sort of by accident part of "Move non-register-requirements-related functionality out of the TreeNodeInfoInit methods into the first phase of Lowering"

Yes, and I'm afraid it will also conflict with the `Lowering` work that is currently being done for arm32. I don't think there's a way to embark on this kind of refactoring and restructuring without stepping on each others' toes.
And I'm not sure I would really call it "by accident" - it's an area ripe for refactoring!

**Comment by pgavlin** [comment]
> Sounds good. My old experiment placed ALL the nodes in a SmallHashTable.

Ah, yeah--that probably wouldn't work out so well without some tweaks to the policy used for resizing. When the number of nodes in a `SmallHashTable` is large, it will probably be overloaded under the current scheme.
